vsphere/internal/virtualdevice/virtual_machine_disk_subresource.go

```diff
@@ -0,0 +1,1514 @@
+	device, err := r.FindVirtualDevice(l)
+	if err != nil {
+		return nil, fmt.Errorf("cannot find disk device: %s", err)
+	}
+	disk, ok := device.(*types.VirtualDisk)
+	if !ok {
+		return nil, fmt.Errorf("device at %q is not a virtual disk", l.Name(device))
+	}
+	deleteSpec, err := object.VirtualDeviceList{disk}.ConfigSpec(types.VirtualDeviceConfigSpecOperationRemove)
+	if err != nil {
+		return nil, err
+	}
+	if r.Get("keep_on_remove").(bool) || r.Get("attach").(bool) {
+		// Clear file operation so that the disk is kept on remove.
+		deleteSpec[0].GetVirtualDeviceConfigSpec().FileOperation = ""
+	}
+	log.Printf("[DEBUG] %s: Device config operations from update: %s", r, DeviceChangeString(deleteSpec))
+	log.Printf("[DEBUG] %s: Delete completed", r)
+	return deleteSpec, nil
+}
+
+// ValidateDiff performs any complex validation of an individual disk
+// sub-resource that can't be done in schema alone.
+//
+// Do not use resourceData in this function as it's not populated and calls to
+// it will cause a panic.
+func (r *DiskSubresource) ValidateDiff() error {
+	log.Printf("[DEBUG] %s: Beginning diff validation (device information may be incomplete)", r)
+	name := r.Get("name").(string)
+
+	// Enforce the maximum unit number, which is the current value of
+	// scsi_controller_count * 15 - 1.
+	ctlrCount := r.resourceDiff.Get("scsi_controller_count").(int)
+	maxUnit := ctlrCount*15 - 1
+	currentUnit := r.Get("unit_number").(int)
+	if currentUnit > maxUnit {
+		return fmt.Errorf("unit_number on disk %q too high (%d) - maximum value is %d with %d SCSI controller(s)", name, currentUnit, maxUnit, ctlrCount)
+	}
+
+	switch r.Get("attach").(bool) {
+	case true:
+		switch {
+		case r.Get("datastore_id").(string) == "":
+			return fmt.Errorf("datastore_id for disk %q is required when attach is set", name)
+		case r.Get("size").(int) > 0:
+			return fmt.Errorf("size for disk %q cannot be defined when attach is set", name)
+		case r.Get("eagerly_scrub").(bool):
+			return fmt.Errorf("eagerly_scrub for disk %q cannot be defined when attach is set", name)
+		case r.Get("keep_on_remove").(bool):
+			return fmt.Errorf("keep_on_remove for disk %q is implicit when attach is set, please remove this setting", name)
+		}
+	default:
+		if r.Get("size").(int) < 1 {
+			return fmt.Errorf("size for disk %q: required option not set", name)
+		}
+	}
+	log.Printf("[DEBUG] %s: Diff validation complete", r)
+	return nil
+}
+
+// NormalizeDiff checks the diff for a vsphere_virtual_machine disk
+// sub-resource. It should be called after setting data and olddata to values
+// that seems similar enough to be possibly the same resource, after which this
+// function should then compare them and fill in any values that may be missing
+// from the new set due to computed values, such as inferred datastore or path
+// names.
+//
+// Do not use resourceData in this function as it's not populated and calls to
+// it will cause a panic.
+func (r *DiskSubresource) NormalizeDiff() error {
```

Comment:
We still need the subresource object here, which is where `resourceData` and `resourceDiff` lay. The API here is indeed a tad unsafe, as conversely `resourceDiff` won't be available in all non-diff calls.

This could be an interface if the `Set` call in `SetRestart` didn't exist in `virtual_machine_device_subresource.go`. I think this could still be an interface - `SetRestart` could type-assert (not necessarily a habit I want to get into 100%) and just silently fail if it doesn't have access to `ResourceData`. Better yet, `SetRestart` could probably just use `SetNew` if it currently has access to `ResourceDiff` instead of `ResourceData`.

I'll make this change.